server/driver/wivrn_foveation.cpp

```diff
@@ -0,0 +1,279 @@
+		                                                 .codeSize = spirv.size() * sizeof(uint32_t),
+		                                                 .pCode = spirv.data(),
+		                                         });
+
+		pipeline = vk::raii::Pipeline(vk.device, nullptr, vk::ComputePipelineCreateInfo{
+		                                                          .stage = {
+		                                                                  .stage = vk::ShaderStageFlagBits::eCompute,
+		                                                                  .module = *shader,
+		                                                                  .pName = "main",
+		                                                          },
+		                                                          .layout = *layout,
+		                                                  });
+	}
+
+	std::array<vk::DescriptorSetLayout, 2> layouts = {*ds_layout, *ds_layout};
+
+	auto my_ds = vk.device.allocateDescriptorSets({
+	        .descriptorPool = *dp,
+	        .descriptorSetCount = 2,
+	        .pSetLayouts = layouts.data(),
+	});
+
+	ds = {
+	        my_ds[0].release(),
+	        my_ds[1].release(),
+	};
+}
+
+void wivrn_foveation_renderer::render_distortion_images(std::array<to_headset::foveation_parameter, 2> foveation_arr, const VkImage * images, const VkImageView * image_views)
+{
+	cmd_buf.reset();
+	cmd_buf.begin(vk::CommandBufferBeginInfo{});
+
+	// Monado's distortion shader is patched to only consider the red channel's image
+	for (int eye = 0; eye < 2; ++eye)
+	{
+		// Descriptor set
+
+		vk::DescriptorImageInfo image_info{
+		        .imageView = image_views[eye],
+		        .imageLayout = vk::ImageLayout::eGeneral,
+		};
+		vk.device.updateDescriptorSets(
+		        vk::WriteDescriptorSet{
+		                .dstSet = ds[eye],
+		                .dstBinding = 0,
+		                .descriptorCount = 1,
+		                .descriptorType = vk::DescriptorType::eStorageImage,
+		                .pImageInfo = &image_info,
+		        },
+		        nullptr);
+
+		std::array im_barriers0 = {
+		        vk::ImageMemoryBarrier{
+		                .srcAccessMask = vk::AccessFlagBits::eNone,
+		                .dstAccessMask = vk::AccessFlagBits::eMemoryWrite,
+		                .oldLayout = vk::ImageLayout::eShaderReadOnlyOptimal,
+		                .newLayout = vk::ImageLayout::eGeneral,
+		                .image = images[eye],
+		                .subresourceRange = {.aspectMask = vk::ImageAspectFlagBits::eColor,
+		                                     .baseMipLevel = 0,
+		                                     .levelCount = 1,
+		                                     .baseArrayLayer = 0,
+		                                     .layerCount = 1},
+		        },
+		};
+		cmd_buf.pipelineBarrier(
+		        vk::PipelineStageFlagBits::eTopOfPipe,
+		        vk::PipelineStageFlagBits::eComputeShader,
+		        {},
+		        nullptr,
+		        nullptr,
+		        im_barriers0);
```

Comment:
I am not sure if the driver is smart enough to do this efficiently or if we should have barriers out of the loop, but with both images